Unify ttnn.div and ttnn.divide

Aswinmcw opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/28936

### Problem description
ttnn have 2 API's for division - div and divide both have different args which leads to confusion

### What's changed
- Removed 'accurate_mode' flag;  added 'fast_and_approximate_mode' across various test files and implementations to unify the 2 API's
- Adjusted related test cases and documentation to reflect the new parameter naming.
- Ensured consistency in handling division operations in both Python and C++ implementations.

This change enhances code readability and aligns with the intended functionality of the division operations.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
https://github.com/tenstorrent/tt-metal/actions/runs/20422034341
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

Aswinmcw committed:
Refactor division function to use 'approx_mode' instead of 'fast_and_approximate_mode'

- Updated the division function calls across multiple files to replace 'fast_and_approximate_mode' with 'approx_mode' for consistency.
- Adjusted related test cases to reflect this change in the division operation.

- [x] All tests pass
- [x] Code review completed

rebase conflicts

Aswinmcw committed:
Refactor division operations to use std::nullopt for optional parameters

### Changes made
- Updated division function calls to replace std::optional<bool> with std::nullopt for clarity.
- Simplified the attachment of golden functions by ensuring that ttnn.divide is set as an alias for ttnn.div after importing operations.
- Adjusted comments to clarify the relationship between ttnn.divide and ttnn.div.

### Checklist
- [x] All tests pass
- [x] Code review completed

Aswinmcw committed:
Refactor division operations to use 'approx_mode' instead of 'accurate_mode'

### Changes made:
- Updated parameter names from 'accurate_mode' to 'approx_mode' across multiple test files and implementations.
- Adjusted related logic in the division functions to accommodate the new parameter name.
- Ensured consistency in test cases and examples to reflect the change in parameter naming.

This refactor aims to clarify the intent of the mode being set for division operations, enhancing code readability and maintainability.

Aswinmcw committed:
Refine legacy argument handling in ExecuteDiv operation

### Changes made:
- Updated the logic for determining legacy argument usage in the `ExecuteDiv::invoke` method.
- Clarified comments to specify that only `round_mode` and `approx_mode=true` trigger the legacy path, while `output_dtype` can be used with both legacy and new paths.
- Enhanced error messages for better clarity regarding the validity of parameters when using the legacy option.

This update improves code readability and ensures more accurate error handling in division operations.

Aswinmcw committed:
Refactor division operations to use 'fast_and_approximate_mode' instead of 'approx_mode'

### Changes made:
- Updated the division functions in `tt_encoder.py`, `tt_oftnet.py`, and various test files to replace 'approx_mode' with 'fast_and_approximate_mode' for consistency and clarity.
- Adjusted parameter names in tests and examples to reflect the new naming convention.
- Ensured that all relevant tests were updated to maintain functionality.

This change enhances the readability of the code and aligns with the updated API specifications.

Aswinmcw committed:
Update division operation to default to 'fast_and_approximate_mode' and clean up related code

### Changes made:
- Set the default value of 'fast_and_approximate_mode' to `true` in the division bindings and related functions.
- Removed outdated comments and unnecessary bindings for the golden function in `binary_complex.py` and `binary.py`.
- Added an alias for the division operation in `__init__.py` for easier access.

This update enhances the performance of division operations while maintaining backward compatibility.

mcw-anasuya commented on ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_composite_op.cpp:
```diff
@@ -271,12 +271,13 @@ Tensor ExecuteDiv::invoke(
     const bool is_int32 = input_dtype == DataType::INT32 && input_b.dtype() == DataType::INT32;
 
-    const auto has_legacy_only_args = (round_mode.has_value() and !is_int32) or output_dtype.has_value() or accurate_mode;
+    // Only round_mode and fast_and_approximate_mode=true force the legacy path
```
Small note: please re-check which cases are supported in binary ng path.

Aswinmcw committed:
Refine division operation error messages and legacy argument handling

### Changes made:
- Updated error messages for clarity regarding the use of 'round_mode' and 'fast_and_approximate_mode' with the 'use_legacy' parameter in the division operation.
- Adjusted the logic for determining legacy argument conditions to simplify the checks related to 'fast_and_approximate_mode'.

This improves the user experience by providing more precise feedback on invalid parameter combinations.

Aswinmcw committed:
Update legacy path condition in ExecuteDiv to include non-INT32 types for round_mode. This change ensures proper handling of legacy arguments in binary operations.

Aswinmcw committed:
Refine legacy argument handling in ExecuteDiv by removing fast_and_approximate_mode from the condition. This update clarifies the requirements for legacy arguments in binary operations, ensuring compatibility with non-INT32 types for round_mode.

Aswinmcw committed:
Enhance legacy argument validation in ExecuteDiv by incorporating fast_and_approximate_mode into the condition. Update error messages for clarity regarding valid configurations when use_legacy is false, ensuring better handling of non-INT32 types for round_mode and accurate mode settings.

Aswinmcw committed:
Refine legacy argument validation in ExecuteDiv by removing the dependency on fast_and_approximate_mode. Update error messages for clarity regarding the use of round_mode when use_legacy is false, ensuring better understanding of limitations for scalar operations.

Aswinmcw committed:
Refine legacy argument validation in ExecuteDiv by removing the dependency on FLOAT32 types. Update comments and error messages for improved clarity regarding the use of fast_and_approximate_mode and round_mode with non-INT32 types.

KalaivaniMCW commented on ttnn/cpp/ttnn/operations/eltwise/binary/binary_pybind.cpp:
```diff
@@ -2530,9 +2518,11 @@ void py_module(py::module& module) {
         R"doc(
         With INT32 inputs, round_mode `None` produces a FLOAT32 output, while `floor` and `trunc` produce an INT32 output.
+        When :attr:`fast_and_approximate_mode` is `True`, operation assumes that :attr:`input_tensor_b` is not zero for fast approximation.
+        When :attr:`fast_and_approximate_mode` is `False` (default), operation properly handles division by zero (accurate mode).
```
@mcw-anasuya 
When the inputs are INT32, div-by-zero is not handled right ?  output datatype conversion is not supported. ?

KalaivaniMCW committed:
#28936: use BinaryOperationWithFastApprox for div

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/div_cleanup

Aswinmcw committed:
Update binary_nanobind.cpp to rename 'accurate_mode' to 'fast_and_approximate_mode' for clarity and consistency. Adjusted related function signatures and documentation to reflect the new parameter name. Removed outdated comments regarding fast approximation handling.

Aswinmcw committed:
Enhance documentation in binary_nanobind.cpp to clarify behavior of division operations under different rounding modes. Added details on handling division by zero and introduced UINT16 as a valid input type.

Aswinmcw committed:
restore pybind

Aswinmcw committed:
Refactor division operations in binary and ternary computations to use floor mode consistently. Updated calls to `ttnn::div` to set the rounding mode to false across multiple files, ensuring uniform behavior in remainder and fmod operations.

KalaivaniMCW committed:
#28936: verify use_legacy, mem_config usage

KalaivaniMCW committed:
#28936: Unify Div TS variant and Divide TS variant

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/div_cleanup

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/div_cleanup

Aswinmcw commented:
/ping @dchenTT @tenstorrent/metalium-developers-ttnn-core

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/div_cleanup

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/div_cleanup

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/div_cleanup

dchenTT commented on ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_composite_op.cpp:
```diff
@@ -200,53 +200,110 @@ Tensor ExecuteDiv::invoke(
+        } else {
+            // round_mode = None
+            TT_FATAL(
+                (output_dtype == std::nullopt || output_dtype == DataType::FLOAT32),
```
!output_dtype.has_value() is more readable.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/div_cleanup

Aswinmcw committed:
Refactor condition checks in ExecuteDiv::invoke for clarity

Updated condition checks for `use_legacy` and `output_dtype` in the `ExecuteDiv::invoke` method to use `value_or` and `has_value` for improved readability and to eliminate potential nullopt issues. This change enhances the clarity of the error messages related to integer division and output data type validation.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/div_cleanup
